### PR TITLE
Add new parameter to prevent grub-mkconfig being run

### LINKED ISSUE #47: Can the call to grub2-mkconfig be disableable?
I use augeasproviders_grub to configure flash card images that contain a configured operating system including /etc/default/grub but that do not have grub or /boot/grub2/grub.cfg yet.  That gets installed at burn time when the OS image is actually written to disk.  

Would it be possible to add an option to disable the execution of the grub2-mkconfig step so that I can have it skipped when running in my build containers?

Thank you

### PR DESCRIPTION
**Pull request by cedws**
Closes #47.
I'm afraid I don't really know Ruby tooling so I am not sure if the tests pass.

### CODE REVIEW

File: lib/puppet/type/grub_user.rb [hunk 1]
```diff
@@ -78,6 +78,13 @@
     end
   end
 
+  newparam(:mkconfig, :boolean => true) do
```
Comment (raphink) [inline lib/puppet/type/grub_user.rb:81]: I think I'd rather call this `run_mkconfig` to be clearer.

File: lib/puppet/type/kernel_parameter.rb [hunk 2]
```diff
@@ -32,6 +32,13 @@
     defaultto :all
   end
 
+  newparam(:mkconfig, :boolean => true) do
```
Comment (raphink) [inline lib/puppet/type/kernel_parameter.rb:35]: Same here, let's call it `run_mkconfig`

File: lib/puppet/provider/grub_config/grub2.rb [hunk 3]
```diff
@@ -101,6 +101,9 @@ def flush
     fail("Cannot find grub.cfg location to use with grub-mkconfig") unless cfg
 
     super
-    mkconfig "-o", cfg
+
+    if mkconfig
+      mkconfig "-o", cfg
+    end
```
Comment (raphink) [inline lib/puppet/provider/grub_config/grub2.rb:107]: Parameters are retrieved with `resource[:param]`, so

```suggestion

    mkconfig "-o", cfg if resource[:run_mkconfig]
```

should do

File: lib/puppet/provider/grub_user/grub2.rb [hunk 4]
```diff
@@ -284,7 +284,9 @@ def flush
       FileUtils.chmod(0755, resource[:target])
     end
 
-    mkconfig "-o", grub2_cfg_path
+    if mkconfig
+      mkconfig "-o", grub2_cfg_path
+    end
```
Comment (raphink) [inline lib/puppet/provider/grub_user/grub2.rb:289]: Same here:

```suggestion
    mkconfig "-o", grub2_cfg_path if resource[:run_mkconfig]
```

**Comment by raphink** [comment]
This looks good. Now it only needs some tests :smile: 


Something similar to https://github.com/hercules-team/augeasproviders_grub/blob/master/spec/unit/puppet/provider/kernel_parameter/grub2_spec.rb#L73, but it should expect to never run when `run_mkconfig => false`.

**Comment by cedws** [comment]
Apologies for the delay on this. I had a look at writing some tests but am having some toolchain issues. I've installed the dependencies with `bundler` and am trying to run the tests with `rspec`, but it's spitting out some [errors](https://gist.github.com/cedws/a309ff1b25d83e69d76fd88423f172d8). Any ideas?

**Comment by raphink** [comment]
Are you running `bundle exec rake spec`?

**Comment by cedws** [comment]
Ah, no, I was just running `rspec` from GEMPATH, thanks. I've just committed a single test, is this along the right lines? I assume you would like me to add a test for the other types as well (`grub_config`, `grub_user`, et al)?

**Comment by raphink** [comment]
Yes, please!

**Comment by cedws** [comment]
Sorry, not getting very far with this. Tried running `bundle exec rake acceptance` on Ruby 2.7 but get [this](https://gist.github.com/cedws/8b2eb2b6e898084bbbd42bc2b557fb3d). Then I saw CI might be using 2.4.4 (though it doesn't appear to actually run the acceptance tests?) so I downgraded, now `bundle install` is [failing](https://gist.github.com/cedws/d0fd6a2598a1a20486d323e9b0269bc6) because it looks like something to do with Augeas is failing to compile. I have the headers installed on my system.
Also tried doing this all in Docker but I guess the acceptance tests require virtualisation so that won't work.
I do need to add acceptance tests for the `run_mkconfig` parameter for the `grub_menuentry`/`grub_user`/`...` types correct?

**Comment by raphink** [comment]
There's no acceptance tests in this module, all the tests are declared as unit tests, so you run all the tests with `bundle exec rake spec`.

**Comment by trevor-vaughan** [comment]
There actually are acceptance tests and it would be great to run them. The CI can't run them because they require virtual machines and are run through `beaker`.

I suggest using PDK to set up an isolated dev environment or RVM or rbenv but not system Ruby.

**Comment by raphink** [comment]
Oh my bad, I didn't think we had any acceptance tests in augeasproviders modules. Sorry about that. 

And I'm guessing docker containers would not be sufficient for grub tests.

**Comment by trevor-vaughan** [comment]
Nope, docker won't work at all for GRUB. Need the real deal :-D.

I had to add them due to the complexity of GRUB itself and just being too dangerous to get wrong.

**Comment by raphink** [comment]
We used to run beaker on Travis with openstack or AWS nodes, but it's a paid solution.

**Comment by trevor-vaughan** [comment]
Yeah, that's the issue. Without someone to put up a system to run the tests, there's not much that can be done besides run them by hand.

**Comment by trevor-vaughan** [comment]
There actually are acceptance tests and it would be great to run them. The CI can't run them because they require virtual machines and are run through `beaker`.

I suggest using PDK to set up an isolated dev environment or RVM or rbenv but not system Ruby.

**Comment by trevor-vaughan** [comment]
I'll be able to review this next week if it can wait that long.

**Comment by trevor-vaughan** [comment]
I'm getting failures in beaker with the code that I'll dig into tomorrow.

**Comment by trevor-vaughan** [comment]
@cedws Let us know if you need help with the testing. Things are 90% there!

**Comment by cedws** [comment]
@trevor-vaughan Thanks very much for this. I will not be able to get to it for a week or so.

**Comment by jcpunk** [comment]
I'll confess interest in this patch...

**Comment by raphink** [comment]
@cedws are you still interested in working on this?

**Comment by cedws** [comment]
I am, but have quite a lot of things going on at the moment. It's still on my to-do list. Anybody is welcome to pick this up.

**Pull request closed without merging**